README.md
# Microparse

This library contains the microparse to c compiler's source code.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2021-07-25
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2021-07-25
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/c-fsm-byte-gen/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/c-fsm-byte-gen/compare/0.1.0...0.1.1


c-code-template/test.c
/*@ requires \valid(v+(0..n-1));
 requires n > 0;
 assigns v[0..n-1];
 ensures \forall integer q; 0<=q<=n-1 ==> v[q]==(unsigned char)0; */
static void make_zero( unsigned char *v, int n ) {

  unsigned char *p = (unsigned char*)v;

  /*@
    loop invariant 0 <= n <= \at(n, Pre);
    loop invariant p == v+(\at(n, Pre)-n);
    loop invariant \forall integer j;  0 <= j < (\at(n, Pre)-n) ==> v[j] == (unsigned char)0;
    loop assigns n, p, v[0..\at(n, Pre)-n-1];
    loop variant n;
  */

  while( n-- ){
    *p++ = 0;
  }
}


test/c_fsm_byte_gen/core_test.clj
(ns c-fsm-byte-gen.core-test
  (:require [clojure.test :refer :all]
            [c-fsm-byte-gen.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


project.clj
(defproject c-fsm-byte-gen "0.1.0-SNAPSHOT"
  :description "byte-based FSM generator"
  :url "http://example.com/FIXME"
  :license {:name "EPL-2.0 OR GPL-2.0-or-later WITH Classpath-exception-2.0"
            :url "https://www.eclipse.org/legal/epl-2.0/"}
  :dependencies [[org.clojure/clojure "1.10.1"]
                 [ubergraph "0.8.2"]
                 [com.taoensso/timbre "5.1.0"]
                 [metasoarous/oz "1.6.0-alpha31"]
                 [macroz/tangle "0.2.2"]
                 [jcf/dorothy "0.0.7-SNAPSHOT"]
                 [walmartlabs/datascope "0.1.1"]
                 [de.ubercode.clostache/clostache "1.4.0"]
                 [org.clojure/math.numeric-tower "0.0.4"]]
  :main c-fsm-byte-gen.core
  :plugins [[lein-auto "0.1.3"]]
  :repl-options {:init-ns c-fsm-byte-gen.core}
  :profiles {:paravim {:dependencies [[paravim "RELEASE"]]
                   :main paravim.start
                   ; on mac os, an additional jvm arg must be added
                   ; so opengl can work correctly
                   :jvm-opts ~(if (= "Mac OS X" (System/getProperty "os.name"))
                                ["-XstartOnFirstThread"] [])}})


c-code-template/main.c
#include <stdio.h>
#include <stdbool.h>
#include <assert.h>

#define INF -1
#define NINF -2

// functiona callback
// typedef void (*FunctionCallback)(int);
// FunctionCallback grammar_constructs[] = { &tag_cons, &len_cons, &case_cons, &repeat_cons };

typedef enum {
    BIT,
    BYTE
} dataunit;

typedef enum {
    LSB,
    MSB
} endianness;

typedef enum {
    TAG,
    LEN,
    REPEAT,
    CASE
} cons_t;

typedef struct {
    int start_byte;
    int start_bit;
    int end_byte;
    int end_bit;
} parsed_result;

int cur_byte_pos = 0;
int cur_bit_pos = 0;

int save_head_byte_pos;
int save_head_bit_pos;

const char* TAPE = "\xF2\xFF\xFF"; 
const int TAPE_LEN = 3;


/*@requires TAPE_LEN >= 0;
  requires \valid_read(((char*)TAPE)+(0..TAPE_LEN));
  requires cur_byte_pos < TAPE_LEN;
  requires cur_byte_pos >= 0;
  requires cur_bit_pos >= 0;
  requires cur_bit_pos <= 31;
  
  assigns cur_byte_pos, cur_bit_pos;
  
  ensures cur_bit_pos >= 0 && cur_bit_pos <= 8;
  ensures cur_byte_pos >= 0 && cur_byte_pos <= TAPE_LEN;
*/
bool get_next_bit() {
    bool cur_symbol;

    // assert(cur_byte_pos < TAPE_LEN);
    char cur_char = TAPE[cur_byte_pos];
    //@assert cur_bit_pos < 32;
    cur_symbol = 1 & (cur_char >> cur_bit_pos);
    // printf("Read %d from tape at loc %d\n", cur_symbol, cur_byte_pos*8+cur_bit_pos);
    //@assert cur_byte_pos + 1 < 2147483647;
    cur_bit_pos++;
    if (cur_bit_pos >= 8) {
        cur_bit_pos = cur_bit_pos % 8;
        cur_byte_pos++;
    }

    return cur_symbol;
}


/*@requires TAPE_LEN >= 0;
  requires \valid_read(TAPE+(0..(TAPE_LEN-1)));
  requires cur_byte_pos <= TAPE_LEN;
  requires cur_byte_pos >= 0;
  requires cur_bit_pos >= 0;
  requires cur_bit_pos <= 8;
  assigns \nothing;
  ensures (\result == 0) || (\result == 1);
*/
bool is_tape_available() {
    if ((cur_byte_pos + 1) < TAPE_LEN) {
        // some bytes remaining
        return 1;
    } else {
        if ((cur_byte_pos + 1) == TAPE_LEN) {
            // some bits left
            if (cur_bit_pos < 7) {
                return 1;
            }
        }
    }
    return 0;
}

/*@requires \valid(r);
  requires cur_bit_pos >= 0 && cur_bit_pos <= 8;
  requires cur_byte_pos >= 0 && cur_byte_pos <= TAPE_LEN;
  assigns *r;
*/
void set_pos_end(parsed_result* r)
{
    r->end_bit  = cur_bit_pos;
    r->end_byte = cur_byte_pos;
    return;
}

/*@requires \valid(r);
  requires \separated(r, &cur_byte_pos, &cur_bit_pos);
  requires cur_bit_pos >= 0 && cur_bit_pos <= 8;
  requires cur_byte_pos >= 0 && cur_byte_pos <= TAPE_LEN;
  assigns *r;
*/
void set_pos_start(parsed_result* r)
{
    r->start_bit  = cur_bit_pos;
    r->start_byte = cur_byte_pos;
    return;
}


/*@requires TAPE_LEN >= 0;
  requires \valid_read(((char*)TAPE)+(0..TAPE_LEN));
  requires cur_byte_pos < TAPE_LEN;
  requires cur_byte_pos >= 0;
  requires cur_bit_pos >= 0;
  requires cur_bit_pos < 8;
  
  assigns cur_byte_pos;
  
  ensures cur_bit_pos >= 0 && cur_bit_pos <= 8;
  ensures cur_byte_pos >= 0 && cur_byte_pos <= TAPE_LEN;
*/
unsigned char get_next_byte() {
    char cur_char = TAPE[cur_byte_pos];
    // printf("Read %d from tape at loc %d\n", cur_char, cur_byte_pos);
    cur_byte_pos++;
    /*@assert cur_byte_pos >= 0;*/
    return cur_char;
}

/*@requires TAPE_LEN >= 0;
  requires cur_byte_pos <= TAPE_LEN;
  requires cur_byte_pos >= 0;
  requires cur_bit_pos >= 0;
  requires cur_bit_pos <= 8;
  assigns \nothing;
  ensures \result >= 0;
*/
int remaining_bits() {
    int remaining_bytes = TAPE_LEN - cur_byte_pos;
    int remaining_bits;
    if (cur_bit_pos == 0) {
        remaining_bits = 0;
    } else {
        remaining_bits = 8 - cur_bit_pos;
    }

    return (8 * remaining_bytes) + remaining_bits;
}

/*@requires TAPE_LEN >= 0;
  requires cur_byte_pos <= TAPE_LEN;
  requires cur_byte_pos >= 0;
  requires cur_bit_pos >= 0;
  requires cur_bit_pos <= 8;
  assigns \nothing;
  ensures \result >= 0;
*/
int remaining_bytes() {
    int remaining_bytes = TAPE_LEN - cur_byte_pos;
    return remaining_bytes;
}

// set the value of the register to T
/*@requires size > 0;
  requires cur_byte_pos <= TAPE_LEN;
  requires cur_byte_pos >= 0;
  requires cur_bit_pos >= 0;
  requires cur_bit_pos <= 8;
  requires ((unit == BIT) ==> size <= 32) && ((unit == BYTE) ==> size <= 4);
  assigns \nothing;
*/
int tag_cons(dataunit unit, int size) {
    unsigned int tag_val;

    if (unit == BIT) {
        /*@ loop invariant 0 <= i < size;
          loop assigns tag_val;
          loop variant size;*/
        for (int i = 0; i < size; ++i) {
            tag_val = tag_val | get_next_bit();
            tag_val = tag_val << 1;
        }
        //@ assert tag_val >= 0;
    } else {
        assert(size <= 4);
        /*@ loop invariant 0 <= i < size;
          loop assigns tag_val;
          loop variant size;*/
        for (int i = 0; i < size; ++i) {
            tag_val = get_next_byte();
            tag_val = tag_val << 8;
        }
        //@ assert tag_val >= 0;
    }
    return tag_val;
};

// set the value of the counter register to L.
/*@requires TAPE_LEN >= 0;
  requires \valid_read(TAPE+(0..(TAPE_LEN-1)));
  requires cur_byte_pos <= TAPE_LEN;
  requires cur_byte_pos >= 0;
  requires cur_bit_pos >= 0;
  requires cur_bit_pos <= 8;
  requires size > 0;
  requires ((unit == BIT) ==> size <= 32) && ((unit == BYTE) ==> size <= 4);
  assigns \nothing;
*/
int len_cons(dataunit unit, endianness e, int size) {
    int len_val = 0;
    if (unit == BIT) {
        int i = 0;
        /*@loop invariant 0 <= i < size;
          loop assigns len_val;
          loop variant size;*/
        for (i = 0; i < size; ++i) {
            len_val = len_val | get_next_bit();
            len_val = len_val << 1;
        }
        //@ assert len_val >= 0;
    } else {
        int i = 0;
        // assert(size <= 4);
        //@ requires size <= 4;
        //@ assert unit == BYTE;
        /*@loop invariant 0 <= i < size;
          loop assigns len_val;
          loop variant size;*/
        for (i = 0; i < size; ++i) {
            len_val = get_next_byte();
            len_val = len_val << 8;
        }
        //@ assert len_val >= 0;
    }

    // if (unit == BIT) {
    //     if (remaining_bits() < len_val) {
    //         // error
    //         debug_write("len: err\n");
    //     }
    // } else {
    //     if (remaining_bytes() < len_val) {
    //         debug_write("len:err\n");
    //     }
    // }
    // return read value
    return len_val;
};

// make lookup table and call that function.
// case function: generate this with code
// repeats construct r, n times

/*@requires TAPE_LEN >= 0;
  requires \valid_read(TAPE+(0..(TAPE_LEN-1)));
  requires cur_byte_pos <= TAPE_LEN;
  requires cur_byte_pos >= 0;
  requires cur_bit_pos >= 0;
  requires cur_bit_pos <= 8;
  requires n > 0 || min > 0 || max > 0;
  assigns \nothing;
*/
parsed_result repeat_cons(dataunit u, int n, int min, int max) {
    parsed_result r;
    set_pos_start(&r);
    if (n == INF) {
        // at least min items and at most max items
        if (u == BIT) {
            int i = 0;
            /*@ loop invariant  0 <= i <= min;
               loop assigns \nothing;
               loop variant min;*/
            for (i = 0; i < min; ++i) {
                get_next_bit();
            }
            /*@ loop invariant  0 <= i <= max;
               loop assigns \nothing;
               loop variant max;*/
            while (i <= max) {
                if (is_tape_available()) {
                    break;
                }
                get_next_bit();
                i++;
            }

            set_pos_end(&r);
            return r;
        } else {
            int i = 0;

            /*@ loop invariant  0 <= i <= min;
               loop assigns \nothing;
               loop variant min;*/
            for (i = 0; i < min; ++i) {
                get_next_byte();
            }

            /*@ loop invariant  0 <= i <= max;
               loop assigns \nothing;
               loop variant max;*/
            while (i <= max) {
                if (is_tape_available()) {
                    break;
                }
                get_next_byte();
            }

            set_pos_end(&r);
            return r;
        }
    } else {
        if (u == BIT) {
            assert(n <= 8);
            /*@loop invariant  0 <= i < n;
              loop assigns \nothing;
              loop variant n;*/
            for (int i = 0; i < n; ++i) {
                get_next_bit();
            }
            set_pos_end(&r);
            return r;
        } else {
            /*@loop invariant  0 <= i < n;
               loop assigns \nothing;
               loop variant n;*/
            for (int i = 0; i < n; ++i) {
                get_next_byte();
            }
            set_pos_end(&r);
            return r;
        }
    }
};


// We define a gloabl list of registers which store results.
/*@ assigns \nothing;
*/
unsigned int run_parser(char* input_buffer, int buffer_len) {
    // PDU
    int reg4144 = tag_cons(BIT, 4);
    // RFU
    int reg4145 = tag_cons(BIT, 1);
    // ChSel
    int reg4146 = tag_cons(BIT, 1);
    // TxAdd
    int reg4147 = tag_cons(BIT, 1);
    // RxAdd
    int reg4148 = tag_cons(BIT, 1);
    int reg4149 = len_cons(BYTE, LSB, 2);

    switch (reg4144) {
        case 0b0001: {
                         // :ADV_DIRECT_IND
                         // AdvA
                         parsed_result  reg4150 = repeat_cons(BYTE, 6, NINF, INF);
                         // Target A
                         parsed_result  reg4151 = repeat_cons(BYTE, 6, NINF, INF);
                         break;
                     }
        case 0b0011: {
                         // :SCAN_REQ
                         // ScanA
                         parsed_result  reg4152 = repeat_cons(BYTE, 6, NINF, INF);
                         // AdvA
                         parsed_result  reg4153 = repeat_cons(BYTE, 6, NINF, INF);
                         break;
                     }
        case 0b0000: {
                     // :ADV_IDV
                     // AdvA
                     parsed_result  reg4154 = repeat_cons(BYTE, 6, NINF, INF);
                     // repeat 0 31 bytes
                     parsed_result  reg4155 = repeat_cons(BYTE, INF, 0, 31);
                     break;
                 }
        case 0b0101: {
                         // :CONNECT_IND
                         // InitA
                         parsed_result  reg4156 = repeat_cons(BYTE, 6, NINF, INF);
                         // AdvA
                         parsed_result  reg4157 = repeat_cons(BYTE, 6, NINF, INF);
                         // AA
                         parsed_result  reg4158 = repeat_cons(BYTE, 4, NINF, INF);
                         // CRCInit
                         parsed_result  reg4159 = repeat_cons(BYTE, 3, NINF, INF);
                         // WinSize
                         parsed_result  reg4160 = repeat_cons(BYTE, 1, NINF, INF);
                         // WinOffset
                         parsed_result  reg4161 = repeat_cons(BYTE, 2, NINF, INF);
                         // Interval
                         parsed_result  reg4162 = repeat_cons(BYTE, 2, NINF, INF);
                         // Latency
                         parsed_result  reg4163 = repeat_cons(BYTE, 2, NINF, INF);
                         // Timeout
                         parsed_result  reg4164 = repeat_cons(BYTE, 2, NINF, INF);
                         // ChM
                         parsed_result  reg4165 = repeat_cons(BYTE, 5, NINF, INF);
                         // Hop
                         parsed_result  reg4166 = repeat_cons(BYTE, 5, NINF, INF);
                         // SCA
                         parsed_result  reg4167 = repeat_cons(BYTE, 3, NINF, INF);
                         break;
                     }
        case 0b0010: {
                         // :ADV_NONCONN_IND
                         // AdvA
                         parsed_result  reg4168 = repeat_cons(BYTE, 6, NINF, INF);
                         // repeat 0 31 bytes
                         parsed_result  reg4169 = repeat_cons(BYTE, INF, 0, 31);
                         break;
                     }
        case 0b0111: {
                         // :ADV_EXT_IND
                         int reg4170 = len_cons(BIT, LSB, 6);
                         // AdvMode
                         parsed_result  reg4171 = repeat_cons(BYTE, 2, NINF, INF);
                         if (reg4170 == 0) {} else {
                             // repeat 0 63 bytes
                             parsed_result  reg4172 = repeat_cons(BYTE, reg4170, 0, 63);
                         }
                         // repeat 0 - 254 bytes
                         parsed_result  reg4173 = repeat_cons(BYTE, INF, 0, 254);
                         break;
                     }
        case 0b1000: {
                         // :AUX_CONNECT_RSP
                         int reg4174 = len_cons(BIT, LSB, 6);
                         // AdvMode
                         parsed_result  reg4175 = repeat_cons(BYTE, 2, NINF, INF);
                         if (reg4174 == 0) {} else {
                             // repeat 0 63 bytes
                             parsed_result  reg4176 = repeat_cons(BYTE, reg4174, 0, 63);
                         }
                         // repeat 0 - 254 bytes
                         parsed_result  reg4177 = repeat_cons(BYTE, INF, 0, 254);
                         break;
                     }
        case 0b0100: {
                         // :SCAN_RSP
                         // AdvA
                         parsed_result  reg4178 = repeat_cons(BYTE, 6, NINF, INF);
                         // repeat 0 31 bytes
                         parsed_result  reg4179 = repeat_cons(BYTE, INF, 0, 31);
                         break;
                     }
        case 0b0110: {
                         // :ADV_SCAN_IND
                         // AdvA
                         parsed_result  reg4180 = repeat_cons(BYTE, 6, NINF, INF);
                         // repeat 0 31 bytes
                         parsed_result  reg4181 = repeat_cons(BYTE, INF, 0, 31);
                         break;
                     }
        default:
            //printf("ERROR: None matched case expression!");
            return 1;
            break;
    }
    return 0;
};

/*@assigns \nothing; */
int main(int argc, char *argv[])
{
    unsigned int x = run_parser(TAPE, TAPE_LEN);
    if (x == 0) {
        return 0;
        // printf("accepted input.");
    } else {
        // printf("rejected input");
        return 1;
    }
    return 0;
}


src/c_fsm_byte_gen/core.clj
(ns c-fsm-byte-gen.core
  (:require [clojure.string :as str])
  (:require [clojure.pprint :as pp])
  (:require [tangle.core :as t])
  (:require [com.walmartlabs.datascope :as ds])
  (:require [taoensso.timbre :as timbre
             :refer [log  trace  debug  info  warn  error  fatal  report
                     logf tracef debugf infof warnf errorf fatalf reportf
                     spy get-env]]))

(def global-reg-cons-dict (atom {}))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; Constructs
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
(defn gen-tag
  "Generates a tag"
  [unit size dict desc
   & {:keys [ref-len-counter]}]
  {:type :tag
   :unit unit
   :dict dict
   :size size
   :ref-len-counter ref-len-counter
   :id (gensym "tag")
   :desc desc})

(defn gen-len
  "Generates a length"
  [unit endianness size desc
   & {:keys [min-size]}]
  {:type :length
   :unit unit
   :endianness endianness
   :id (gensym "len")
   :size size
   :min-size min-size
   :desc desc})

(defn gen-repeat
  "Generates a repeat construct"
  [unit repeat-times desc
   & {:keys [range-min range-max ref-len-counter]
      :or {range-min ##-Inf range-max ##Inf ref-len-counter nil}}]
  {:type :repeat
   :unit unit
   :repeat-times repeat-times
   :range-min range-min
   :range-max range-max
   :ref-len-counter ref-len-counter
   :id (gensym "repeat")
   :desc desc})

(defn gen-case
  "Generates a case construct"
  [grammar-cons case-dict
   & {:keys [ref-len-counter]
      :or {ref-len-counter nil}}]
  {:type :case
   :grammar-cons grammar-cons
   :ref-len-counter ref-len-counter
   :id (gensym "case")
   :case-dict case-dict})

(defn gen-start-counter "" [reg]
  {:type :counter :reg reg :id (gensym "counter")})

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; Grammar Definitions Begins Here
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
(def PDU
  (gen-tag
    :bit
    4
    {:ADV_IND         "0000"
     :ADV_DIRECT_IND  "0001"
     :ADV_NONCONN_IND "0010"
     :SCAN_REQ        "0011"
     :SCAN_RSP        "0100"
     :CONNECT_IND     "0101"
     :ADV_SCAN_IND    "0110"
     :ADV_EXT_IND     "0111"
     :AUX_CONNECT_RSP "1000"}
    "PDU"))

(def RFU   (gen-tag :bit 1 {:RFU_ON    "1" :RFU_OFF    "0"} "RFU"))
(def ChSel (gen-tag :bit 1 {:CHSEL_ON  "1" :CHSEL_OFF  "0"} "ChSel"))
(def TxAdd (gen-tag :bit 1 {:TX_ADD_ON "1" :TX_ADD_OFF "0"} "TxAdd"))
(def RxAdd (gen-tag :bit 1 {:RX_ADD_ON "1" :RX_ADD_OFF "0"} "RxAdd"))

(def repeat-bytes
  (gen-repeat :byte ##Inf "repeat * bytes"))

(def repeat-0-31-bytes
  (gen-repeat :byte ##Inf "repeat 0 31 bytes"
              :range-min 0
              :range-max 31))

(def ADV_IND [(gen-repeat :byte 6 "AdvA")
              repeat-0-31-bytes])

(def ADV_DIRECT_IND [(gen-repeat :byte 6 "AdvA")
                     (gen-repeat :byte 6 "Target A")])

(def ADV_NONCONN_IND [(gen-repeat :byte 6 "AdvA")
                      repeat-0-31-bytes])

(def SCAN_REQ [(gen-repeat :byte 6 "ScanA")
               (gen-repeat :byte 6 "AdvA")])

(def SCAN_RSP [(gen-repeat :byte 6 "AdvA")
               repeat-0-31-bytes])

(def LLDATA [(gen-repeat :byte 4 "AA")
             (gen-repeat :byte 3 "CRCInit")
             (gen-repeat :byte 1 "WinSize")
             (gen-repeat :byte 2 "WinOffset")
             (gen-repeat :byte 2 "Interval")
             (gen-repeat :byte 2 "Latency")
             (gen-repeat :byte 2 "Timeout")
             (gen-repeat :byte 5 "ChM")
             (gen-repeat :byte 5 "Hop")
             (gen-repeat :byte 3 "SCA")])

(def CONNECT_IND (concat [(gen-repeat :byte 6 "InitA")
                          (gen-repeat :byte 6 "AdvA")]
                       LLDATA))

(def ADV_SCAN_IND [(gen-repeat :byte 6 "AdvA")
                   repeat-0-31-bytes])

(def Extended-Header-Length
  (gen-len :bit :LSB 6 "Extended Header Length"))

(def AdvData
  (gen-repeat :byte ##Inf "repeat 0 - 254 bytes" :range-min 0 :range-max 254))

(def ExtendedHeader
  (gen-repeat :byte ##Inf "repeat 0 63 bytes" :range-min 0 :range-max 63))

(def Common-Extended-Advertising-Payload
  [Extended-Header-Length
   (gen-repeat :byte 2 "AdvMode")
   (gen-case Extended-Header-Length
             {:if (list 0 nil)
              :else ExtendedHeader})
  AdvData])

(def ADV_EXT_IND Common-Extended-Advertising-Payload)

(def AUX_CONNECT_RSP Common-Extended-Advertising-Payload)

(def payload-grammar
  (gen-case PDU
            {:ADV_IDV ADV_IND
             :ADV_DIRECT_IND  ADV_DIRECT_IND
             :ADV_NONCONN_IND ADV_NONCONN_IND
             :SCAN_REQ        SCAN_REQ
             :SCAN_RSP        SCAN_RSP
             :CONNECT_IND     CONNECT_IND
             :ADV_SCAN_IND    ADV_SCAN_IND
             :ADV_EXT_IND     ADV_EXT_IND
             :AUX_CONNECT_RSP AUX_CONNECT_RSP}))

(def len-field
  (gen-len :byte :LSB 2 "header"))

(def packet-grammar
  [PDU RFU ChSel TxAdd RxAdd len-field payload-grammar])

(def output-file-handle
  (clojure.java.io/writer "foo.txt"))

(defn gen-register "return a register name" []
  (gensym "reg"))

(defn convert-tag-to-c-code "doc-string" [tag-cons]
  "converts a tag construct to c code"
  (let [unit  (case (:unit tag-cons) :bit "BIT" :byte "BYTE")
        size (:size tag-cons)
        reg (gen-register)]
    (swap! global-reg-cons-dict conj {(:id tag-cons) reg})
    (str "// " (:desc tag-cons) " \n"
      "int " reg " = " "tag_cons(" unit ", " size ");")))

(defn convert-length-to-c-code "doc-string" [len-cons]
  (let [unit  (case (:unit len-cons) :bit "BIT" :byte "BYTE")
        endianness (case (:endianness len-cons) :LSB "LSB" :MSB "MSB")
        size (:size len-cons)
        reg (gen-register)]
    (swap! global-reg-cons-dict conj {(:id len-cons) reg})
    (str "int " reg " = "
         "len_cons(" unit ", "  endianness ", " size ");\n")))

(declare convert-cons-to-c-code)
(defn convert-range-inf-to-c "doc-string" [a]
  (case a ##Inf "INF" ##-Inf "NINF" a))

(defn convert-repeat-to-c-code "doc-string" [repeat-cons]
  (let [reg (gen-register)
        unit (clojure.string/upper-case (name (:unit repeat-cons)))]
    (str "// " (:desc repeat-cons) "\n"
      "parsed_result  " reg " = repeat_cons(" unit ", "
      (convert-range-inf-to-c (:repeat-times repeat-cons)) ", "
      (convert-range-inf-to-c (:range-min repeat-cons)) ", "
      (convert-range-inf-to-c (:range-max repeat-cons)) ");\n")))

(defn convert-tag-case-to-c-code "doc-string" [case-cons]
  (let [predicate-g (:grammar-cons case-cons)
        predicate-g-dict (:dict predicate-g)
        reg-name ((:id predicate-g) @global-reg-cons-dict)
        case-dict (:case-dict case-cons)
        generate-inner-case-code (fn [v]
                                   (apply str
                                     (interpose
                                       "\t"
                                       (map convert-cons-to-c-code v))))
        generate-case-code (fn [[k v]]
                             (let [case-val (k predicate-g-dict)
                                   code-statements (generate-inner-case-code v)]
                               (str "case 0b" case-val ": {\n\t"
                                    "// " (str k) "\n\t"
                                    code-statements
                                    " break;\n}\n")))
        generate-case-statements (map generate-case-code case-dict)
        case-statements (apply str generate-case-statements)]
    (str "switch (" reg-name ") { \n"
         case-statements
         "default:\n\t printf(\"ERROR: None matched case expression!\");\n }")))

(defn convert-len-case-to-c-code "" [case-cons]
  (let [predicate-g (:grammar-cons case-cons)
        reg-name ((:id predicate-g) @global-reg-cons-dict)
        case-dict (:case-dict case-cons)
        if-clause (:if case-dict)
        else-clause (:else case-dict)]
    (str "if (" reg-name " == " (first if-clause) ") {"
        (if (nil? (second if-clause)) ""
          (convert-cons-to-c-code (second if-clause)))
        "} else {\n"
        (apply str (convert-cons-to-c-code
                     (update-in else-clause
                                [:repeat-times]
                                (fn [x] reg-name))))
        "}\n")))

(defn convert-case-to-c-code "doc-string" [case-cons]
  (case (:type (:grammar-cons case-cons))
    :tag (convert-tag-case-to-c-code case-cons)
    :length (convert-len-case-to-c-code case-cons)))

(defn convert-cons-to-c-code "doc-string" [grammar-construct]
  (case (:type grammar-construct)
    :tag  (convert-tag-to-c-code grammar-construct)
    :length (convert-length-to-c-code grammar-construct)
    :repeat (convert-repeat-to-c-code grammar-construct)
    :case (convert-case-to-c-code grammar-construct)))

(defn generate-c-code
  "generates c code given a grammar"
  [grammar]
  (map convert-cons-to-c-code grammar))

;; A lang with following commands:
;; read tag, read len, start_len, repeat (for loop), if (r0 == "") { g0 } else {g1}

(def generated-code
  (apply str (interpose "\n" (generate-c-code packet-grammar))))

(defn -main
  "main func"
  []
  (println "Started...")
  (spit output-file-handle generated-code)
  (println "printing complete..."))
